File: HeadsetPresenter_Bluetools/HeadsetPresenter/Bluetools/BlueTools SDK v1.21/dotNetCF200/C#/ObexPushClient/MainForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.Windows.Forms;
using System.Data;
using System.IO;

using Franson.BlueTools;
using Franson.Protocols.Obex;
using Franson.Protocols.Obex.ObjectPush;

namespace ObjectPushSample200CF
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class MainForm : System.Windows.Forms.Form
	{
		private System.Windows.Forms.ProgressBar progressBar;
		private System.Windows.Forms.Button sendBtn;
		private System.Windows.Forms.Button discoverBtn;
		private System.Windows.Forms.Label label1;
		private System.Windows.Forms.ListBox deviceList;
		private System.Windows.Forms.OpenFileDialog openFileDialog;
		private System.Windows.Forms.Label informLabel;
		private System.Windows.Forms.StatusBar statusBar;

		private Manager m_manager = null;
		private Network m_network = null;
		private ObexObjectPush m_objectPush = null;
		private bool m_boolAborted = false;
		private bool m_boolDenied  = false;
		
		// these objects should be class global to prevent them from being taken by the garbage collector
		private RemoteService m_serviceCurrent	= null;
		private Stream m_streamCurrent			= null;
		private bool m_boolUnrecoverableError   = false;

		public MainForm()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();
			
			// catch all BlueTools exceptions - like e.g. license expired
			try
			{
				// get your trial license or buy BlueTools at franson.com
				Franson.BlueTools.License license = new Franson.BlueTools.License();
				license.LicenseKey = "HU5UZexzz2JOHX68pijhSkaRWu00NRMbHY5X";

				// get bluetools manager
				m_manager = Manager.GetManager();

				// get first network dongle - bluetools 1.0 only supports one!
				m_network = m_manager.Networks[0];

				// marshal events into this class thread.
				m_manager.Parent = this;

				statusBar.Text = GetStack();
			}
			catch (Exception exc)
			{
				// display BlueTools errors here
				MessageBox.Show(exc.Message);
			}
		}
		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			// BlueTools manager must be disposed, otherwise you can't exit application!
			Manager.GetManager().Dispose();

			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.progressBar = new System.Windows.Forms.ProgressBar();
            this.sendBtn = new System.Windows.Forms.Button();
            this.discoverBtn = new System.Windows.Forms.Button();
            this.label1 = new System.Windows.Forms.Label();
            this.deviceList = new System.Windows.Forms.ListBox();
            this.openFileDialog = new System.Windows.Forms.OpenFileDialog();
            this.statusBar = new System.Windows.Forms.StatusBar();
            this.informLabel = new System.Windows.Forms.Label();
            this.SuspendLayout();
            // 
            // progressBar
            // 
            this.progressBar.Location = new System.Drawing.Point(40, 152);
            this.progressBar.Name = "progressBar";
            this.progressBar.Size = new System.Drawing.Size(160, 8);
            // 
            // sendBtn
            // 
            this.sendBtn.Enabled = false;
            this.sendBtn.Location = new System.Drawing.Point(128, 168);
            this.sendBtn.Name = "sendBtn";
            this.sendBtn.Size = new System.Drawing.Size(72, 20);
            this.sendBtn.TabIndex = 3;
            this.sendBtn.Text = "Push";
            this.sendBtn.Click += new System.EventHandler(this.sendBtn_Click);
            // 
            // discoverBtn
            // 
            this.discoverBtn.Location = new System.Drawing.Point(40, 168);
            this.discoverBtn.Name = "discoverBtn";
            this.discoverBtn.Size = new System.Drawing.Size(72, 20);
            this.discoverBtn.TabIndex = 4;
            this.discoverBtn.Text = "Discover";
            this.discoverBtn.Click += new System.EventHandler(this.discoverBtn_Click);
            // 
            // label1
            // 
            this.label1.Location = new System.Drawing.Point(40, 8);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(100, 16);
            this.label1.Text = "Devices";
            // 
            // deviceList
            // 
            this.deviceList.Location = new System.Drawing.Point(40, 32);
            this.deviceList.Name = "deviceList";
            this.deviceList.Size = new System.Drawing.Size(160, 114);
            this.deviceList.TabIndex = 6;
            this.deviceList.SelectedIndexChanged += new System.EventHandler(this.deviceList_SelectedIndexChanged);
            // 
            // statusBar
            // 
            this.statusBar.Location = new System.Drawing.Point(0, 272);
            this.statusBar.Name = "statusBar";
            this.statusBar.Size = new System.Drawing.Size(240, 22);
            // 
            // informLabel
            // 
            this.informLabel.Location = new System.Drawing.Point(8, 200);
            this.informLabel.Name = "informLabel";
            this.informLabel.Size = new System.Drawing.Size(224, 40);
            this.informLabel.Text = "Click on discover to detect Bluetooth devices.";
            // 
            // MainForm
            // 
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Inherit;
            this.ClientSize = new System.Drawing.Size(240, 294);
            this.Controls.Add(this.informLabel);
            this.Controls.Add(this.statusBar);
            this.Controls.Add(this.progressBar);
            this.Controls.Add(this.sendBtn);
            this.Controls.Add(this.discoverBtn);
            this.Controls.Add(this.label1);
            this.Controls.Add(this.deviceList);
            this.MaximizeBox = false;
            this.MinimizeBox = false;
            this.Name = "MainForm";
            this.Text = "Object Push Sample CF";
            this.Closing += new System.ComponentModel.CancelEventHandler(this.MainForm_Closing);
            this.ResumeLayout(false);

		}
		#endregion

		private void discoverBtn_Click(object sender, System.EventArgs e)
		{
			m_network.DeviceDiscovered			+= new BlueToolsEventHandler(m_network_DeviceDiscovered);
			m_network.DeviceDiscoveryStarted	+= new BlueToolsEventHandler(m_network_DeviceDiscoveryStarted);
			m_network.DeviceDiscoveryCompleted	+= new BlueToolsEventHandler(m_network_DeviceDiscoveryCompleted);

			// Start looking for devices on the network
			// Use Network.DiscoverDevices() if you don't want to use events.
			m_network.DiscoverDevicesAsync();
		}

		private void m_network_DeviceDiscovered(object sender, BlueToolsEventArgs eventArgs)
		{
			// add every device found
			RemoteDevice device = (RemoteDevice)((DiscoveryEventArgs)eventArgs).Discovery;
			deviceList.Items.Add(device);
		}

		private void m_network_DeviceDiscoveryStarted(object sender, BlueToolsEventArgs eventArgs)
		{
			// disable the Discover button when we begin device discovery
			discoverBtn.Enabled = false;

			// inform the user what we're doing
			informLabel.Text = "Scanning for devices...";

			// disable device list while scanning
			deviceList.Enabled = false;
		}

		private void m_network_DeviceDiscoveryCompleted(object sender, BlueToolsEventArgs eventArgs)
		{
			// enable Discover button again since device discovery is complete
			discoverBtn.Enabled = true;

			// inform the user what to do
			informLabel.Text = "Click on the device you wish to push a file to.";

			// enable device list again
			deviceList.Enabled = true;

			m_network.DeviceDiscovered			-= new BlueToolsEventHandler(m_network_DeviceDiscovered);
			m_network.DeviceDiscoveryStarted	-= new BlueToolsEventHandler(m_network_DeviceDiscoveryStarted);
			m_network.DeviceDiscoveryCompleted	-= new BlueToolsEventHandler(m_network_DeviceDiscoveryCompleted);
		}

		private void deviceList_SelectedIndexChanged(object sender, System.EventArgs e)
		{
			// cancel any possible ongoing device discovery if user clicks on anything
			m_network.CancelDeviceDiscovery();

			// get selected item (a remote device)
			RemoteDevice selectedDevice = (RemoteDevice)deviceList.SelectedItem;

			selectedDevice.ServiceDiscoveryStarted += new BlueToolsEventHandler(selectedDevice_ServiceDiscoveryStarted);
			selectedDevice.ServiceDiscoveryCompleted += new BlueToolsEventHandler(selectedDevice_ServiceDiscoveryCompleted);
			// detect Object Push service (OPP) on this device
			selectedDevice.DiscoverServicesAsync(ServiceType.OBEXObjectPush);
		}

		private void selectedDevice_ServiceDiscoveryStarted(object sender, BlueToolsEventArgs eventArgs)
		{
			// when beginning to search for services - disable the Discover button
			// we shouldn't scan for devices while attempting to scan for services.
			discoverBtn.Enabled = false;
			sendBtn.Enabled		= false;
		}

		private void selectedDevice_ServiceDiscoveryCompleted(object sender, BlueToolsEventArgs eventArgs)
		{
			// when service discovery is complete - let us re-enable the discover button. 
			// It is okay to try to update the device list
			discoverBtn.Enabled = true;

			// set error event handler for the device
			RemoteDevice device = (RemoteDevice)sender;
			device.Error += new BlueToolsEventHandler(device_Error);

			// get all services found
			Service[] services = (Service[])((DiscoveryEventArgs)eventArgs).Discovery;

			// if we have a service since before, close the stream
			if (m_streamCurrent != null)
			{
				m_streamCurrent.Close();
				m_streamCurrent = null;
			}
			
			// and remove the service
			if (m_serviceCurrent != null)
			{
				m_serviceCurrent = null;
			}

			// if we found a new service...
			if (services.Length > 0)
			{				
				// ...get OPP service object
				m_serviceCurrent = (RemoteService)services[0];
								
				try
				{			
					// create an ObexObjectPush object connected to the ServiceStream
					// wait for 20 seconds for frozen devices
					m_objectPush = new ObexObjectPush(-1);

					// marshal event to this class thread
					m_objectPush.Parent = this;

					// setup event handlers
					m_objectPush.Error				+= new ObexEventHandler(m_objectPush_Error);
					m_objectPush.PutFileBegin		+= new ObexEventHandler(m_objectPush_PutFileBegin);
					m_objectPush.PutFileProgress	+= new ObexEventHandler(m_objectPush_PutFileProgress);
					m_objectPush.PutFileEnd			+= new ObexEventHandler(m_objectPush_PutFileEnd);
                    m_objectPush.DisconnectEnd      += new ObexEventHandler(m_objectPush_DisconnectEnd);

					sendBtn.Enabled		= true;

					informLabel.Text = "Click on Push to select file(s) to push to device.";
				}
				catch (Exception exc)
				{
					sendBtn.Enabled		= false;

					MessageBox.Show(exc.Message);
				}
			}			
		}

		private void sendBtn_Click(object sender, System.EventArgs e)
		{
			if (sendBtn.Text.Equals("Push"))
			{
				try
				{
					// if we have a service to send to
					if (m_serviceCurrent != null)
					{
						// make sure ObexObjectPush is initialized and select file to push to device
						if (m_objectPush != null && openFileDialog.ShowDialog() == DialogResult.OK)
						{
							sendBtn.Text = "Cancel";

							m_streamCurrent = m_serviceCurrent.Stream;

							FileStream fileStream = new FileStream(openFileDialog.FileName, FileMode.Open, FileAccess.Read);
							m_objectPush.PushFileAsync(fileStream, Path.GetFileName(openFileDialog.FileName), m_streamCurrent);
						}
					}
				}
				catch (Exception exc)
				{
					MessageBox.Show(exc.Message);
				}
			}
			else
			{
				// if button text isn't Send it will be Cancel - 
				// if user click it, we tell ObexObjectPush that we want to abort
				m_objectPush.Abort();

				// only need to press cancel once - won't help you to press it more
				sendBtn.Enabled = false;
			}
		}

		private void m_objectPush_PutFileProgress(object sender, ObexEventArgs eventArgs)
		{
			ObexCopyEventArgs copyArgs = (ObexCopyEventArgs)eventArgs;

			// if max value on progressbar isn't set, set it
			if (copyArgs.Size != progressBar.Maximum)
            {
				progressBar.Maximum = (int)copyArgs.Size;
            }

			// set position of file copy
			progressBar.Value = (int)copyArgs.Position;

			// update copy panel to show progress in kb
			statusBar.Text = Convert.ToString(copyArgs.Position / 1024) + " kb / " + Convert.ToString(copyArgs.Size / 1024) + " kb";
		}

		private void m_objectPush_PutFileEnd(object sender, ObexEventArgs eventArgs)
		{
			// when finished copying...

			// ...close file stream
			ObexCopyEventArgs copyArgs = (ObexCopyEventArgs)eventArgs;

			// Close Stream opened by ObexObjectPush
			copyArgs.Stream.Close();

			// ...enable discover button
			discoverBtn.Enabled = true;

			// ...change text caption back to Push
			sendBtn.Text = "Push";

			// set progressbar position back to beginning - just for show, doesn't really matter
			progressBar.Value = 0;

			// set copy panel to nothing to make it invisible again - it's auto-resizing :)
			statusBar.Text = GetStack();

			// inform the user what is happening
			if (m_boolAborted)
			{
				// ...if transfer was aborted..
				informLabel.Text = "Push aborted.";
			}
			else if (m_boolDenied)
			{
				// ... if transfer was impossible
				informLabel.Text = "Push denied by device.";
			}
			else if (m_boolUnrecoverableError)
			{
				informLabel.Text = "Stream was lost to device.";
			}
			else
			{
				// ...transfer completed
				informLabel.Text = "File(s) sent to device.";
			}

            // enable device list again when finished
            deviceList.Enabled = true;

			// show user what to do next
			informLabel.Text += "\nClick on Send to select file(s) to push to device.";

			// we enable the Send button again if it was disabled when it was a Cancel button
			sendBtn.Enabled = true;
		}

		private void m_objectPush_PutFileBegin(object sender, ObexEventArgs eventArgs)
		{
			// this transfer hasn't been aborted (yet!)
			m_boolAborted = false;

			// this transfer hasn't been denied (yet!)
			m_boolDenied = false;

			// this transfer has not had any errors (yet!)
			m_boolUnrecoverableError = false;

			// while copying, disable the discover button
			discoverBtn.Enabled = false;

            // disable device list while copying
            deviceList.Enabled = false;

			// inform the user what is happening
			informLabel.Text = "Sending file(s) to device...";
		}

		private void m_objectPush_Error(object sender, ObexEventArgs eventArgs)
		{
			ObexErrorEventArgs errorArgs = (ObexErrorEventArgs)eventArgs;
			
			// if the error was not an Obex error - we can assume the Stream is not available anymore
			if (errorArgs.Error == Franson.Protocols.Obex.ErrorCode.StreamError)
			{
				// since the stream was lost, re-enable the UI

				// the Push button
				sendBtn.Enabled = false;

				// the device list
				deviceList.Enabled = true;

				// this error can't be recovered from
				m_boolUnrecoverableError = true;

				// show this error in a box - it's pretty serious
				MessageBox.Show(errorArgs.Message);
			}
			else 
			{
				// if the error is transfer was aborted, set boolean so we can display this to user
				if (errorArgs.Error == Franson.Protocols.Obex.ErrorCode.Aborted ||
				    errorArgs.Error == Franson.Protocols.Obex.ErrorCode.SecurityAbort)
                {
					m_boolAborted = true;
				}

				if (errorArgs.Error == Franson.Protocols.Obex.ErrorCode.SendDenied)
				{
					m_boolDenied = true;
				}

				// show message to user
				MessageBox.Show(errorArgs.Message);
			}
		}

		private void device_Error(object sender, BlueToolsEventArgs eventArgs)
		{
			Franson.BlueTools.ErrorEventArgs errorArgs = (Franson.BlueTools.ErrorEventArgs)eventArgs;

			statusBar.Text = errorArgs.Message;
		}

        private void m_objectPush_DisconnectEnd(object sender, ObexEventArgs eventArgs)
        {
            // we close this stream to ensure that we get a fresh new one when pushing next time.
            // this is needed since some devices do not allow multiple pushes in one stream
            if (m_streamCurrent != null)
            {
                m_streamCurrent.Close();
            }
        }

		private string GetStack()
		{
			// update statusbar panel with name of currently used stack
			switch (Manager.StackID)
			{
				case StackID.STACK_MICROSOFT: 
				{
					return "Microsoft Stack";
				}

				case StackID.STACK_WIDCOMM:
				{
					return "Widcomm Stack";
				}

				default:
				{
					return "Unknown stack";
				}
			}
		}

		private void MainForm_Closing(object sender, System.ComponentModel.CancelEventArgs e)
		{
			// if m_serviceCurrent is available, close the stream
			if (m_serviceCurrent != null)
			{
				if (m_streamCurrent != null)
					m_streamCurrent.Close();

				// set these objects to null to mark them for GC
				m_streamCurrent  = null;
				m_serviceCurrent = null;
			}

			// dispose objects - also makes sure that Manager are disposed.
			Dispose();
		}
	}
}

File: HeadsetPresenter_Bluetools/HeadsetPresenter/Bluetools/BlueTools SDK v1.2/dotNetCF/VB.NET/ObexPushClient/MainForm.vb
Imports Franson.BlueTools
Imports System.IO
Imports Franson.Protocols.Obex.ObjectPush
Imports Franson.Protocols.Obex

Public Class MainForm
    Inherits System.Windows.Forms.Form
    Friend WithEvents statusBar As System.Windows.Forms.StatusBar
    Friend WithEvents progressBar As System.Windows.Forms.ProgressBar
    Friend WithEvents sendBtn As System.Windows.Forms.Button
    Friend WithEvents discoverBtn As System.Windows.Forms.Button
    Friend WithEvents label1 As System.Windows.Forms.Label
    Friend WithEvents openFileDialog As System.Windows.Forms.OpenFileDialog
    Friend WithEvents deviceList As System.Windows.Forms.ListBox
    Friend WithEvents informLabel As System.Windows.Forms.Label

    Dim m_manager As Manager
    Dim WithEvents m_network As Network
    Dim m_serviceCurrent As RemoteService
    Dim WithEvents m_deviceCurrent As RemoteDevice
    Dim WithEvents m_objectPush As ObexObjectPush
    Dim m_streamCurrent As ServiceStream
    Dim WithEvents m_deviceSelected As RemoteDevice

    Dim m_boolAborted As Boolean
    Dim m_boolDenied As Boolean
    Dim m_boolUnrecoverableError As Boolean

#Region " Windows Form Designer generated code "

    Public Sub New()
        MyBase.New()

        'This call is required by the Windows Form Designer.
        InitializeComponent()

        'Add any initialization after the InitializeComponent() call

    End Sub

    'Form overrides dispose to clean up the component list.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)

        MyBase.Dispose(disposing)

        ' dispose Manager so application can exit
        If Not m_manager Is Nothing Then

            m_manager.Dispose()

        End If

    End Sub

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Private Sub InitializeComponent()
        Me.statusBar = New System.Windows.Forms.StatusBar
        Me.progressBar = New System.Windows.Forms.ProgressBar
        Me.sendBtn = New System.Windows.Forms.Button
        Me.discoverBtn = New System.Windows.Forms.Button
        Me.label1 = New System.Windows.Forms.Label
        Me.openFileDialog = New System.Windows.Forms.OpenFileDialog
        Me.deviceList = New System.Windows.Forms.ListBox
        Me.informLabel = New System.Windows.Forms.Label
        '
        'statusBar
        '
        Me.statusBar.Location = New System.Drawing.Point(0, 248)
        Me.statusBar.Size = New System.Drawing.Size(240, 22)
        '
        'progressBar
        '
        Me.progressBar.Location = New System.Drawing.Point(16, 144)
        Me.progressBar.Size = New System.Drawing.Size(208, 8)
        '
        'sendBtn
        '
        Me.sendBtn.Enabled = False
        Me.sendBtn.Location = New System.Drawing.Point(128, 160)
        Me.sendBtn.Text = "Push"
        '
        'discoverBtn
        '
        Me.discoverBtn.Location = New System.Drawing.Point(32, 160)
        Me.discoverBtn.Text = "Discover"
        '
        'label1
        '
        Me.label1.Location = New System.Drawing.Point(14, 8)
        Me.label1.Size = New System.Drawing.Size(100, 16)
        Me.label1.Text = "Devices"
        '
        'deviceList
        '
        Me.deviceList.Location = New System.Drawing.Point(14, 32)
        Me.deviceList.Size = New System.Drawing.Size(210, 106)
        '
        'informLabel
        '
        Me.informLabel.Location = New System.Drawing.Point(32, 184)
        Me.informLabel.Size = New System.Drawing.Size(168, 48)
        Me.informLabel.Text = "Click Discover to find nearby Bluetooth devices."
        '
        'MainForm
        '
        Me.Controls.Add(Me.progressBar)
        Me.Controls.Add(Me.sendBtn)
        Me.Controls.Add(Me.discoverBtn)
        Me.Controls.Add(Me.label1)
        Me.Controls.Add(Me.deviceList)
        Me.Controls.Add(Me.informLabel)
        Me.Controls.Add(Me.statusBar)
        Me.Text = "ObjectPush Sample"

    End Sub

#End Region

    Private Sub discoverBtn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles discoverBtn.Click

        m_network.DiscoverDevicesAsync()

    End Sub

    Private Sub sendBtn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles sendBtn.Click

        If sendBtn.Text.Equals("Push") Then

            Try

                ' if we have a service to send to
                If Not m_serviceCurrent Is Nothing Then

                    ' make sure ObexObjectPush is initialized and select file to push to device
                    If Not m_objectPush Is Nothing And openFileDialog.ShowDialog() = DialogResult.OK Then

                        m_streamCurrent = m_serviceCurrent.Stream

                        Dim fileStream As FileStream

                        fileStream = New FileStream(openFileDialog.FileName, FileMode.Open, FileAccess.Read)
                        m_objectPush.PushFileAsync(fileStream, Path.GetFileName(openFileDialog.FileName), m_streamCurrent)

                    End If
                End If
            Catch

                ' disable Push button if there was an error
                sendBtn.Enabled = False

                informLabel.Text = "Failed to get a socket. Click on the device you wish to push a file to."

            End Try
        Else

            ' if button text isn't Send it will be Cancel - 
            ' if user click it, we tell ObexObjectPush that we want to abort
            m_objectPush.Abort()

            ' only need to press cancel once - won't help you to press it more
            sendBtn.Enabled = False

        End If

    End Sub

    Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        ' catch all BlueTools exceptions - like e.g. license expired
        Try

            ' setup license for bluetools - use license key given
            Dim License As Franson.BlueTools.License
            License = New Franson.BlueTools.License

            License.LicenseKey = "HU5UZer022JOLX98oidhQkbRbvv0NRJbHXEc"

            ' get bluetools manager
            m_manager = Manager.GetManager()

            ' get first network dongle - bluetools 1.0 only supports one!
            m_network = m_manager.Networks(0)

            ' marshal events into this class thread.
            m_manager.Parent = Me

            ' update statusbar panel with name of currently used stack
            statusBar.Text = GetStack(Manager.StackID)

        Catch exc As Exception

            ' display BlueTools errors here
            MessageBox.Show(exc.Message)

        End Try

    End Sub

    Private Function GetStack(ByVal stackID As Int32)

        Select Case stackID

            Case Franson.BlueTools.StackID.STACK_WIDCOMM

                Return "Microsoft Stack"

            Case Franson.BlueTools.StackID.STACK_WIDCOMM

                Return "Widcomm Stack"

            Case Else

                Return "Unknown stack"

        End Select

    End Function

    Private Sub MainForm_Closing(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles MyBase.Closing

        'if currentService is available, close the stream
        If Not m_serviceCurrent Is Nothing Then

            If Not m_streamCurrent Is Nothing Then

                m_streamCurrent.Close()

                ' set these objects to null to mark them for GC
                m_streamCurrent = Nothing
                m_serviceCurrent = Nothing
            End If
        End If

        ' must dispose manager else application can't exit!
        If Not m_manager Is Nothing Then

            m_manager.Dispose()

        End If

    End Sub

    Private Sub m_network_DeviceDiscovered(ByVal sender As Object, ByVal eventArgs As Franson.BlueTools.BlueToolsEventArgs) Handles m_network.DeviceDiscovered

        ' add every device found
        Dim device As RemoteDevice
        Dim discovery As DiscoveryEventArgs
        discovery = DirectCast(eventArgs, DiscoveryEventArgs)
        device = DirectCast(discovery.Discovery, RemoteDevice)
        deviceList.Items.Add(device)

    End Sub

    Private Sub m_network_DeviceDiscoveryStarted(ByVal sender As Object, ByVal eventArgs As Franson.BlueTools.BlueToolsEventArgs) Handles m_network.DeviceDiscoveryStarted

        ' disable the Discover button when we begin device discovery
        discoverBtn.Enabled = False

        ' disable the send button since you can't send while device discovering
        sendBtn.Enabled = False

        ' disable the device list while searching for devices
        deviceList.Enabled = False

        ' inform the user what we're doing
        informLabel.Text = "Scanning for devices..."

    End Sub

    Private Sub m_network_DeviceDiscoveryCompleted(ByVal sender As Object, ByVal eventArgs As Franson.BlueTools.BlueToolsEventArgs) Handles m_network.DeviceDiscoveryCompleted

        Dim discovery As DiscoveryEventArgs
        discovery = DirectCast(eventArgs, DiscoveryEventArgs)

        Dim devices As Device() = DirectCast(discovery.Discovery, Device())

        ' enable Discover button again since device discovery is complete
        discoverBtn.Enabled = True

        ' enable the device list again
        deviceList.Enabled = True

        ' inform the user what to do
        informLabel.Text = "Click on the device you wish to push a file to."

    End Sub

    Private Sub deviceList_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles deviceList.SelectedIndexChanged

        ' get selected item (a remote device)
        m_deviceSelected = DirectCast(deviceList.SelectedItem, RemoteDevice)

        ' detect Object Push service (OPP) on this device

        deviceList.Enabled = False
        m_deviceSelected.DiscoverServicesAsync(ServiceType.OBEXObjectPush)

    End Sub

    Private Sub m_deviceSelected_ServiceDiscoveryStarted(ByVal sender As Object, ByVal eventArgs As Franson.BlueTools.BlueToolsEventArgs) Handles m_deviceSelected.ServiceDiscoveryStarted

        ' when beginning to search for services - disable the Discover button
        ' we shouldn't scan for devices while attempting to scan for services.
        discoverBtn.Enabled = False
        sendBtn.Enabled = False

    End Sub

    Private Sub m_deviceSelected_ServiceDiscoveryCompleted(ByVal sender As Object, ByVal eventArgs As Franson.BlueTools.BlueToolsEventArgs) Handles m_deviceSelected.ServiceDiscoveryCompleted

        ' when service discovery is complete - let us re-enable the discover button. 
        ' It is okay to try to update the device list
        discoverBtn.Enabled = True

        ' re-enable the device list
        deviceList.Enabled = True

        ' set error event handler for the device
        Dim device As RemoteDevice = DirectCast(sender, RemoteDevice)

        ' get all services found
        Dim discovery As DiscoveryEventArgs
        discovery = DirectCast(eventArgs, DiscoveryEventArgs)

        Dim services As Service() = DirectCast(discovery.Discovery, Service())

        ' if we have a service since before, close the stream
        If Not m_streamCurrent Is Nothing Then

            m_streamCurrent.Close()
            m_streamCurrent = Nothing
        End If

        ' and remove the service
        If Not m_serviceCurrent Is Nothing Then

            m_serviceCurrent = Nothing
        End If

        ' if we found a new service...
        If services.Length > 0 Then

            ' ...get OPP service object
            m_serviceCurrent = DirectCast(services(0), RemoteService)

            Try

                ' create an ObexObjectPush object connected to the ServiceStream
                m_objectPush = New ObexObjectPush(-1) ' wait forever

                ' marshal event to this class thread
                m_objectPush.Parent = Me

                ' enable the push button
                sendBtn.Enabled = True

                ' inform the user of the next step
                informLabel.Text = "Click on Push to select file(s) to push to device."

            Catch exc As Exception

                sendBtn.Enabled = False

                MessageBox.Show(exc.Message)

            End Try
        End If

    End Sub

    Private Sub m_objectPush_PutFileBegin(ByVal sender As Object, ByVal e As Franson.Protocols.Obex.ObexEventArgs) Handles m_objectPush.PutFileBegin

        ' this transfer hasn't been aborted (yet!)
        m_boolAborted = False

        'this transfer hasn't been denied (yet!)
        m_boolDenied = False

        ' this transfer has not had any errors (yet!)
        m_boolUnrecoverableError = False

        ' while copying, disable the discover button
        discoverBtn.Enabled = False

        sendBtn.Text = "Cancel"

        ' disable device list while copying
        deviceList.Enabled = False

        ' inform the user what is happening
        informLabel.Text = "Sending file(s) to device..."

    End Sub

    Private Sub m_objectPush_PutFileProgress(ByVal sender As Object, ByVal eventArgs As Franson.Protocols.Obex.ObexEventArgs) Handles m_objectPush.PutFileProgress

        Dim copyArgs As ObexCopyEventArgs = DirectCast(eventArgs, ObexCopyEventArgs)

        ' if size isn't unknown
        If Not copyArgs.Size = -1 Then
            ' if max value on progressbar isn't set, set it
            If Not copyArgs.Size = progressBar.Maximum Then
                progressBar.Maximum = copyArgs.Size
            End If
            ' set position of file copy
            progressBar.Value = copyArgs.Position

            ' update copy panel to show progress in kb
            statusBar.Text = Convert.ToString(copyArgs.Position \ 1024) + " kb / " + Convert.ToString(copyArgs.Size \ 1024) + " kb"
        Else

            ' update copy panel to show progress in kb
            statusBar.Text = Convert.ToString(copyArgs.Position \ 1024) + " kb / ? kb"

        End If

    End Sub

    Private Sub m_objectPush_PutFileEnd(ByVal sender As Object, ByVal eventArgs As Franson.Protocols.Obex.ObexEventArgs) Handles m_objectPush.PutFileEnd

        ' when finished copying...

        ' ...close file stream
        Dim copyArgs As ObexCopyEventArgs = DirectCast(eventArgs, ObexCopyEventArgs)

        ' Close Stream opened by ObexObjectPush
        copyArgs.Stream.Close()

        ' ...enable discover button
        discoverBtn.Enabled = True

        ' ...change text caption back to Send
        sendBtn.Text = "Push"

        ' set progressbar position back to beginning - just for show, doesn't really matter
        progressBar.Value = 0

        ' show stack name again
        statusBar.Text = GetStack(Manager.StackID)

        ' inform the user what is happening
        If m_boolAborted Then

            ' ...if transfer was aborted..
            informLabel.Text = "Push aborted. Click on Push to select file(s) to push to device."

        ElseIf (m_boolDenied) Then

            ' .. if push was denied by receiving device
            informLabel.Text = "Push denied by device. Click on Push to select file(s) to push to device."

        ElseIf (m_boolUnrecoverableError) Then

            ' ..if stream was lost
            informLabel.Text = "Stream lost to device. Click on the device you wish to push file(s) to."

        Else

            ' ...transfer completed
            informLabel.Text = "File(s) sent to device. Click on Push to select file(s) to push to device."
        End If

        ' enable device list again when finished
        deviceList.Enabled = True

        ' we enable the Send button again if it was disabled when it was a Cancel button
        sendBtn.Enabled = True

    End Sub

    Private Sub m_objectPush_Error(ByVal sender As Object, ByVal eventArgs As Franson.Protocols.Obex.ObexEventArgs) Handles m_objectPush.Error

        Dim errorArgs As ObexErrorEventArgs = DirectCast(eventArgs, ObexErrorEventArgs)

        ' if the error was not an Obex error - we can assume the Stream is not available anymore
        If errorArgs.Error = Franson.Protocols.Obex.ErrorCode.StreamError Then

            ' the Push button
            sendBtn.Enabled = False

            ' the device list
            deviceList.Enabled = True

            ' this error can't be recovered from
            m_boolUnrecoverableError = True

            ' show this error in a box - it's pretty serious
            MessageBox.Show(errorArgs.Message)

        Else

            ' if the error is transfer was aborted, set boolean so we can display this to user
            If errorArgs.Error = Franson.Protocols.Obex.ErrorCode.Aborted Then

                m_boolAborted = True

            End If

            If errorArgs.Error = Franson.Protocols.Obex.ErrorCode.SecurityAbort Then

                m_boolAborted = True

            End If

            If errorArgs.Error = Franson.Protocols.Obex.ErrorCode.SendDenied Then

                m_boolDenied = True
            End If

            ' show message to user
            MessageBox.Show(errorArgs.Message)
        End If

    End Sub

    Private Sub m_deviceCurrent_Error(ByVal sender As Object, ByVal eventArgs As Franson.BlueTools.BlueToolsEventArgs) Handles m_deviceCurrent.Error

        Dim errorArgs As Franson.BlueTools.ErrorEventArgs = DirectCast(eventArgs, Franson.BlueTools.ErrorEventArgs)

        statusBar.Text = errorArgs.Message

    End Sub

    Private Sub m_objectPush_DisconnectEnd(ByVal sender As Object, ByVal e As Franson.Protocols.Obex.ObexEventArgs) Handles m_objectPush.DisconnectEnd

        ' we close this stream to ensure that we get a fresh new one when pushing next time.
        ' this is needed since some devices do not allow multiple pushes in one stream
        If Not m_streamCurrent Is Nothing Then

            m_streamCurrent.Close()

        End If

    End Sub
End Class
